File: backoffice/FondaBeckEndLogic/RestaurantsSearch/GetRestaurantsCommand.cs
using System;
using com.ds201625.fonda.DataAccess.InterfaceDAO;
using com.ds201625.fonda.Domain;
using System.Collections.Generic;
using com.ds201625.fonda.BackEndLogic.Exceptions;

namespace com.ds201625.fonda.BackEndLogic
{
	/// <summary>
	/// Comando para obtener una lista de restaurantes
	/// </summary>
	public class GetRestaurantsCommand : BaseCommand
	{
		public GetRestaurantsCommand () : base ()
		{ }

		/// <summary>
		/// Inicializa los parametros.
		/// </summary>
		/// <returns>Los parametros.</returns>
		protected override Parameter[] InitParameters ()
		{
			// recive 3 parametros
			Parameter[] parameter = new Parameter[5];

			// [0] String de busqueda.
			parameter [0] = new Parameter (typeof(string), false);

			// [1] Id de la zona
			parameter [1] = new Parameter (typeof(int), false);

			// [1] Id de la categoria de restaurant
			parameter [2] = new Parameter (typeof(int), false);

			// [1] limite de resultados.
			parameter [3] = new Parameter (typeof(int), false);

			// [2] Pagina de resultados.
			parameter [4] = new Parameter (typeof(int), false);

			return parameter;
		}

		/// <summary>
		/// Ejecucion del parametro.
		/// </summary>
		protected override void Invoke()
		{
			string q = (string) GetParameter (0);
			int max = -1;
			int page = 1;
			int idZone = 0;
			int idRestCat = 0;

			if (GetParameter (1) != null)
				idZone = (int) GetParameter (1);
			if (GetParameter (2) != null)
				idRestCat = (int) GetParameter (2);
			if (GetParameter (3) != null)
				max = (int) GetParameter (3);
			if (GetParameter (4) != null)
				page = (int) GetParameter (4);

			IRestaurantDAO restDao = FacDao.GetRestaurantDAO ();

			IList<Restaurant> restaurants = null;

			try {
				restaurants = restDao.FindByFilters (q, idZone, idRestCat, max, page);
			} 
			catch (Exception e)
			{
				throw new FondaBackendLogicException ("Error al obtener lista de Restaurantes", e);
			}

			Result = restaurants;
		}
	}
}

File: frontoffice/app/src/main/java/com/ds201625/fonda/logic/Commands/RestaurantCommands/GetRestaurantsCommand.java
package com.ds201625.fonda.logic.Commands.RestaurantCommands;

import android.util.Log;
import com.ds201625.fonda.data_access.factory.FondaServiceFactory;
import com.ds201625.fonda.data_access.services.RestaurantService;
import com.ds201625.fonda.domains.Restaurant;
import com.ds201625.fonda.logic.BaseCommand;
import com.ds201625.fonda.logic.CommandInternalErrorException;
import com.ds201625.fonda.logic.Parameter;
import com.ds201625.fonda.logic.ParameterOutOfIndexException;
import com.ds201625.fonda.logic.SessionData;

import java.util.List;

/**
 * Comando para obtener restaurantes
 */
public class GetRestaurantsCommand extends BaseCommand {

    private String TAG = "GetRestaurantsCommand";

    /**
     * Asignacion de los parametros del comando.
     * Parametro en posicion 0: Query
     * Parametro en posicion 1: Max
     * Pareametro en posicion 2: Page
     * Parametro en posicion 3: Zone
     * Parametro en posicion 4: Category
     */
    @Override
    protected Parameter[] setParameters() {
        Parameter [] parameters = new Parameter[5];
        parameters[0] = new Parameter(String.class, false);
        parameters[1] = new Parameter(Integer.class, false);
        parameters[2] = new Parameter(Integer.class, false);
        parameters[3] = new Parameter(Integer.class, false);
        parameters[4] = new Parameter(Integer.class, false);

        return parameters;
    }

    @Override
    protected void invoke() throws Exception {
        Log.d(TAG, "Comando para obtener la lista de restaurantes");
        List<Restaurant> restaurants = null;

        RestaurantService resService = FondaServiceFactory.getInstance()
                .getRestaurantService(SessionData.getInstance().getToken());

        int max = 0;
        int page = 0;
        int zone = 0;
        int category = 0;

        try
        {
            String query = (String) getParameter(0);

            if (getParameter(1) != null) {
                max = (int) getParameter(1);
            }
            if (getParameter(2) != null) {
                page = (int) getParameter(2);
            }
            if (getParameter(3) != null) {
                zone = (int) getParameter(3);
            }
            if (getParameter(4) != null) {
                category = (int) getParameter(4);
            }
            restaurants = resService.getRestaurants(query, max, page, category, zone);
        } catch (ParameterOutOfIndexException e) {
            Log.e("Fonda Command",e.getMessage());
            throw CommandInternalErrorException.generate(this.getClass().toString(),e);
        }

        setResult(restaurants);
    }
}
